### ISSUE TITLE
[Tracking][WebLLM] Function calling and Embeddings

### ISSUE THREAD

**Issue by CharlieFRuan** [mlc-ai/web-llm#526]
This issue tracks various action items we would like to complete with regard to the features function calling and embeddings.

### Function calling

- [ ] **F1** Enable manual function calling
  - That is, function calling with only `system`, `user`, `assistant`, `tool` messages, without using the `tools` and `tool_calls` fields of OpenAI API
  - Reach parity with the examples provided in MLC-LLM: https://github.com/mlc-ai/mlc-llm/pull/2744
  - This requires various runtime changes as initiated in https://github.com/mlc-ai/web-llm/pull/467
  - Add examples for Llama3.1 and Hermes2 after supported
- [ ] **F2** Use `tools` and `tool_calls` to support function calling following how OpenAI API does it
  - Essentially supporting the example in https://platform.openai.com/docs/guides/function-calling.
    - There are new fields in the official OpenAI API, which we should support as well if possible.
  - This may limit the flexibility for the user. For instance, while Llama3.1 offers roughly 3 formats for function calling, using `tools` will force us to use only one of them
  - The previous PR only offers minimal one-round function calling support https://github.com/mlc-ai/web-llm/pull/451
    - We want to allow the model to make tool calls or respond in natural language at its own discretion
- [ ] **F3** Use BNFGrammar to guarantee tool call generation correctness
  - This requires the model to use a special token to signify the beginning of a function call, `<tool_call>` in the case of Hermes2. Upon such a token being generated, we instantiate a BNFGrammar instance. When ended, force it to generate `</tool_call>`. Before and after this tool call, the model can either generate natural language or other tool calls.
  - The previous PR forces the response to be a tool call, which limits the flexibility a lot https://github.com/mlc-ai/web-llm/pull/451

### Embedding

- [ ] **E1** Loading multiple models within an engine
  - For applications like RAG, two models are needed to complete this, one embedding model and one LLM. We'd like to hold all models in a single `MLCEngine` instead of instantiating multiple engines. This makes `MLCEngine` behave like an endpoint, and offers the possibility for intra-engine optimizations in the future.
  - This requires changes to APIs like `reload()`, `unload()` on top of the recent change https://github.com/mlc-ai/web-llm/pull/525
  - We'd also need to enable `ChatCompletionRequest.model`, when multiple models are loaded.
  - We will try our best to make sure current single-model usages are not broken
- [ ] **E2** Fix concurrent request issues
  - With a single model, we encounter correctness issues with multiple concurrent requests as broughtup in https://github.com/mlc-ai/web-llm/issues/522
  - After E1, we need to further pay close attention to potential concurrency issues.
- [ ] **E3** Implement `engine.embeddings.create()`
  - Implement an `src/openai_api_protocols/embedding.ts` for API, parallel to `src/openai_api_protocols/chat_completion.ts`
  - Implement an `EmbeddingPipeline` in `src/embeddings.ts`, parallel to `LLMChatPipeline` in `llm_chat.ts`
  - Hook up the API entry in `MLCEngine`
- [ ] **E4** Add an example for RAG

**Comment by CharlieFRuan** [comment]
Some future TODOs for embeddings:
- [ ]  Support nomic-v1.5 for longer context (may not simply return first token logits like snowflake-arctic, hence requiring changes in `EmbeddingPipeline`)
- [ ]  Support matryoshka models, hence support `dimension` field for those models
- [ ]  Implement prefill chunking (currently not needed since we only support 512 window size)
